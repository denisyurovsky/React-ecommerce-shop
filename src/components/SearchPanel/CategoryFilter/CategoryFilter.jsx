import {
  Select,
  MenuItem,
  FormControl,
  InputLabel,
  CircularProgress,
} from '@mui/material';
import { Box } from '@mui/system';
import PropTypes from 'prop-types';
import React, { useState } from 'react';

import styles from './CategoryFilter.module.scss';

export const CategoryFilter = ({
  allCategories,
  isLoading,
  errorOccurred,
  className,
  setSearchParams,
}) => {
  const [category, setCategory] = useState(allCategories[0]);

  const handleSelectChange = (e) => {
    setCategory(e.target.value);
    const categoryFilter =
      e.target.value === allCategories[0]
        ? null
        : [{ 'category.name': e.target.value }];

    setSearchParams((prev) => ({
      ...prev,
      filters: categoryFilter,
      currentPage: 1,
    }));
  };

  return (
    <FormControl
      disabled={isLoading || errorOccurred ? true : false}
      className={className}
      sx={{ position: 'relative' }}
    >
      {isLoading && (
        <Box className={styles.preloader}>
          <CircularProgress />
        </Box>
      )}
      <InputLabel htmlFor="uncontrolled-native">Category</InputLabel>
      <Select
        sx={{ minWidth: '180px' }}
        data-testid="select-category-input"
        onChange={handleSelectChange}
        value={category}
        label="Category"
        inputProps={{
          name: 'categories',
          id: 'uncontrolled-native',
        }}
      >
        {allCategories.map((item, index) => (
          <MenuItem key={index} value={item}>
            {item}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};

CategoryFilter.propTypes = {
  allCategories: PropTypes.array.isRequired,
  setSearchParams: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
  errorOccurred: PropTypes.bool.isRequired,
  className: PropTypes.string,
};

export default CategoryFilter;
