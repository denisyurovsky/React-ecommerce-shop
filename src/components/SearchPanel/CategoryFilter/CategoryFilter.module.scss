.preloader {
  display: flex;
  position: absolute;
  top: 8px;
  left: 65px;
}
