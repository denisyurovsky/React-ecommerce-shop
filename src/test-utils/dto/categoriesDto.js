const categoriesDto = [
  'All categories',
  'Grocery',
  'Industrial',
  'Clothing',
  'Computers',
];

export default categoriesDto;
