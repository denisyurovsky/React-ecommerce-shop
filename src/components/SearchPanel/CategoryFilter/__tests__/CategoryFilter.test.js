import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import categoriesDto from '../../../../test-utils/dto/categoriesDto';
import CategoryFilter from '../CategoryFilter';

describe('CategoryFilter component', () => {
  describe('snapshots', () => {
    it('renders a valid snapshot', () => {
      const setSearchParams = jest.fn();
      const { asFragment } = render(
        <CategoryFilter
          allCategories={categoriesDto}
          setSearchParams={setSearchParams}
          isLoading={false}
          errorOccurred={false}
        />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('main group of tests', () => {
    it('CategoryFilter works properly', async () => {
      let data;
      const setSearchParams = jest.fn((func) => (data = func()));

      render(
        <CategoryFilter
          allCategories={categoriesDto}
          setSearchParams={setSearchParams}
          isLoading={false}
          errorOccurred={false}
        />
      );

      const button = screen.getByRole('button', { haspopup: 'listbox' });

      userEvent.click(button);
      const option1 = screen.getByRole('option', { name: /computers/i });

      userEvent.click(option1);
      expect(data.filters).toEqual([{ 'category.name': 'Computers' }]);

      userEvent.click(button);
      const option2 = screen.getByRole('option', { name: /all categories/i });

      userEvent.click(option2);
      expect(data.filters).toEqual(null);
    });
  });
});
