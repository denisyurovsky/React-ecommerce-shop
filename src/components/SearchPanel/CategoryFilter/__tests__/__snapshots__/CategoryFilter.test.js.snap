// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryFilter component snapshots renders a valid snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root css-16ojqud-MuiFormControl-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="uncontrolled-native"
    >
      Category
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-11gsl5e-MuiInputBase-root-MuiOutlinedInput-root"
      data-testid="select-category-input"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="mui-component-select-categories"
        class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        id="mui-component-select-categories"
        role="button"
        tabindex="0"
      >
        All categories
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        id="uncontrolled-native"
        name="categories"
        tabindex="-1"
        value="All categories"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1z7n62"
        >
          <span>
            Category
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</DocumentFragment>
`;
